gpflow/actions.py

```diff
@@ -0,0 +1,386 @@
+# Action implementations
+# ======================
+
+class Loop(Action):
+    """
+    Action loop abstraction.
+    Default loop is an infinite with starting point at zero and update step one. 
+    It evaluates body action as many times as it is specified by start, stop and step.
+    Loop provides two control signals - 'Break' and 'Continue'.
+    These signals must be raised as a standard exception. Depending on which signal
+    was raised the loop either stops execution or continues from beginning.
+
+        :param action: Action or list of actions, which will be run within loop.
+        :param stop: Maximum loop iteration.
+        :param start: Starting point for loop iterations.
+        :param step: Integer interval between iterations.
+    """
+
+    class Continue(Exception):
+        """Continue signal returns execution to the beginning of the loop."""
+        pass
+
+    class Break(Exception):
+        """Break signal stops loop immediately."""
+        pass
+
+    def __init__(self,
+            action: Action,
+            stop: Optional[int] = None,
+            start: int = 0,
+            step: int = 1) -> None:
+        self._action = _try_convert_action(action)
+        self.start = start
+        self.stop = stop
+        self.step = step
+    
+    @property
+    def iteration(self) -> int:
+        """Active iteration number."""
+        return _get_attr(self, _iteration=None)
+    
+    def with_iteration(self, iteration: int):
+        """
+        Set iteration number.
+        
+            :param iteration: Iteration interger number.
+            :return: Loop itself.
+        """
+        self._iteration = iteration
+        return self
+
+    def with_action(self, action: Action) -> 'Loop':
+        """
+        Set loop body action.
+            
+            :param action: Loop body action.
+            :return: Loop itself.
+        """
+        self._action = _try_convert_action(action)
+        return self
+
+    def with_settings(self,
+            stop: Optional[int] = None,
+            start: int = 0,
+            step: int = 1) -> 'Loop':
+        """
+        Set start, stop and step loop configuration.
+
+            :param stop: Looop stop iteration integer. If None then loop
+                becomes infinite.
+            :param start: Loop iteration start integer.
+            :param step: Loop iteration interval integer.
+            :return: Loop itself.
+        """
+        self.start = start
+        self.stop = stop
+        self.step = step
+        return self
+
+    def run(self, context: ActionContext):
+        """
+        Run performs loop iterations.
+
+            :param context: Action context.
+        """
+        iterator = itertools.count(start=self.start, step=self.step)
+        for i in iterator:
+            self.with_iteration(i)
+            if self.stop is not None and i >= self.stop:
+                break
+            try:
+                self._action(context)
```

Comment:
I am not a big fan of using exceptions for control flow - something that is rightly considered a bad idea in any other language. 
Because it's Pythonic (i.e. "Guido said we should do that" ;) ) it's not a big deal, and does not seem to be a source of possible confusion in this context...
Could we instead use enumerators? Could the actions just return an enum value saying whether "break" or "continue"?

Reply:
I think this exception business gets nicer anyway if, rather than exposing publicly the Exception types, what gets exposed publicly is the 'continue_loop' and 'break_loop' functions.

Reply:
1. Exceptions are inevitable, because action may return something meaningful in future to make monad-like execution
2. I like that, I will add break_loop & continue_loop methods